x/ccv/provider/keeper/permissionless.go

```diff
@@ -0,0 +1,414 @@
+	}
+	return ConsumerPhase(buf[0]), true
+}
+
+// SetConsumerIdToPhase sets the phase associated with this consumer id
+// TODO (PERMISSIONLESS): use this method when launching and when stopping a chain
+func (k Keeper) SetConsumerIdToPhase(ctx sdk.Context, consumerId string, phase ConsumerPhase) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.ConsumerIdToPhaseKey(consumerId), []byte{byte(phase)})
+}
+
+// DeleteConsumerIdToPhase deletes the phase associated with this consumer id
+func (k Keeper) DeleteConsumerIdToPhase(ctx sdk.Context, consumerId string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.ConsumerIdToPhaseKey(consumerId))
+}
+
+// GetConsumerIdToStopTime returns the stop time associated with the to-be-stopped chain with consumer id
+func (k Keeper) GetConsumerIdToStopTime(ctx sdk.Context, consumerId string) (time.Time, bool) {
+	store := ctx.KVStore(k.storeKey)
+	buf := store.Get(types.ConsumerIdToStopTimeKey(consumerId))
+	if buf == nil {
+		return time.Time{}, false
+	}
+	var time time.Time
+	if err := time.UnmarshalBinary(buf); err != nil {
+		panic(fmt.Errorf("failed to unmarshal time: %w", err))
+	}
+	return time, true
+}
+
+// SetConsumerIdToStopTime sets the stop time associated with this consumer id
+func (k Keeper) SetConsumerIdToStopTime(ctx sdk.Context, consumerId string, stopTime time.Time) {
+	store := ctx.KVStore(k.storeKey)
+	buf, err := stopTime.MarshalBinary()
+	if err != nil {
+		panic(fmt.Errorf("failed to marshal time: %w", err))
+	}
+	store.Set(types.ConsumerIdToStopTimeKey(consumerId), buf)
+}
+
+// DeleteConsumerIdToStopTime deletes the stop time associated with this consumer id
+func (k Keeper) DeleteConsumerIdToStopTime(ctx sdk.Context, consumerId string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.ConsumerIdToStopTimeKey(consumerId))
+}
+
+// GetClientIdToConsumerId returns the consumer id associated with this client id
+func (k Keeper) GetClientIdToConsumerId(ctx sdk.Context, clientId string) (string, bool) {
+	store := ctx.KVStore(k.storeKey)
+	buf := store.Get(types.ClientIdToConsumerIdKey(clientId))
+	if buf == nil {
+		return "", false
+	}
+	return string(buf), true
+}
+
+// SetClientIdToConsumerId sets the client id associated with this consumer id
+func (k Keeper) SetClientIdToConsumerId(ctx sdk.Context, clientId string, consumerId string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.ClientIdToConsumerIdKey(clientId), []byte(consumerId))
+}
+
+// DeleteClientIdToConsumerId deletes the client id to consumer id association
+func (k Keeper) DeleteClientIdToConsumerId(ctx sdk.Context, clientId string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.ClientIdToConsumerIdKey(clientId))
+}
+
+// GetInitializedConsumersReadyToLaunch returns the consumer ids of the pending initialized consumer chains
+// that are ready to launch,  i.e., consumer clients to be created.
+func (k Keeper) GetInitializedConsumersReadyToLaunch(ctx sdk.Context) []string {
```

Comment:
@mpoke This replaces the loop we had over the pending consumer additional proposals to see which ones we can execute. It does not contain `spawnTime` in the key and hence the loop is not `spawnTime`-sorted and the loop does not stop early. This loop executes in `BeginBlockInit` and hence might slow things down.
Let me know if you think it's worth it to include `spawnTime` in the key under some new state.

Similarly in  `GetLaunchedConsumersReadyToStop` below we do not loop by `stopTime`.